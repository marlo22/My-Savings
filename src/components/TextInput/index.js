import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Label, Input, Icon } from 'native-base';

const iconTypes = {
  valid: 'checkmark-circle',
  invalid: 'close-circle'
};

const feedbackColor = {
  valid: 'green',
  invalid: 'red'
}

const FeedbackIcon = styled(Icon)`
  color: ${({ status }) => feedbackColor[status]};
`;

const TextInput = ({ label, validation = {}, itemProps = {}, labelProps = {}, inputProps = {} }) => {
  const { status, message } = validation;
  const iconName = iconTypes[status];

  return (
    <>
      <Label {...labelProps}>
        {label}
      </Label>
      <Input {...inputProps} style={{ border: 'solid 1px red' }} />
      {iconName && <FeedbackIcon name={iconName} status={status} />}
    </>
  );
}

TextInput.propTypes = {
  label: PropTypes.string.isRequired,
  labelProps: PropTypes.object,
  inputProps: PropTypes.object.isRequired,
  validation: PropTypes.shape({
    isValid: PropTypes.bool.isRequired,
    message: PropTypes.string
  })
};

export default TextInput;