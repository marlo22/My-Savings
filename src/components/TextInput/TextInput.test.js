import React from 'react';
import { mount as shallow } from 'enzyme';

import TextInput from './';
jest.spyOn(console, 'error').mockImplementation(() => null);

const label = 'E-mail';
const labelProps = { style: { color: 'red' } };
const inputProps = { disabled: true };
const validation = { status: 'valid' };
const feedbackIconDisplayName = 'Styled(Icon)';

const component = shallow(
  <TextInput
    label={label}
    labelProps={labelProps}
    inputProps={inputProps}
    validation={validation}
  />
);

describe('<TextInput />', () => {
  test('should render without errors', () => {
    expect(component).toBeTruthy();
  });

  test('snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  test('should set passed label', () => {
    const labelComponent = component.find('Label');
    expect(labelComponent.text()).toBe(label);
  });

  test('should set passed labelProps', () => {
    const labelComponent = component.find('Label');
    expect(labelComponent.prop('style')).toContain(labelProps.style);
  });

  test('should set passed inputProps', () => {
    const input = component.find('Input');
    expect(input.prop('disabled')).toBe(inputProps.disabled);
  });

  test('should set "checkmark-circle" icon if validation status is set as valid', () => {
    const feedbackIcon = component.find(feedbackIconDisplayName);
    expect(feedbackIcon.prop('name')).toBe('checkmark-circle');
    expect(feedbackIcon.prop('status')).toBe('valid');
  });

  test('should set "close-circle" icon if validation status is set as valid', () => {
    component.setProps({ validation: { status: 'invalid' } });
    
    const feedbackIcon = component.find(feedbackIconDisplayName);
    expect(feedbackIcon.prop('name')).toBe('close-circle');
    expect(feedbackIcon.prop('status')).toBe('invalid');
  });

  test('should not render icon if validation status is not set', () => {
    component.setProps({ validation: undefined });
    
    const feedbackIcon = component.find(feedbackIconDisplayName);
    expect(feedbackIcon).toHaveLength(0);
  });
});