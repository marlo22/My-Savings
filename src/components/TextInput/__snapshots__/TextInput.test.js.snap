// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> snapshot 1`] = `
<TextInput
  inputProps={
    Object {
      "disabled": true,
    }
  }
  label="E-mail"
  labelProps={
    Object {
      "style": Object {
        "color": "red",
      },
    }
  }
  validation={
    Object {
      "status": "valid",
    }
  }
>
  <Styled(Label)
    style={
      Object {
        "color": "red",
      }
    }
  >
    <Label
      style={
        Array [
          Object {
            "fontSize": 17,
          },
          Object {
            "color": "red",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "color": "red",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 17,
              },
              Object {
                "color": "red",
              },
            ]
          }
        >
          E-mail
        </Text>
      </Text>
    </Label>
  </Styled(Label)>
  <Styled(Input)
    disabled={true}
    style={
      Object {
        "border": "solid 1px red",
      }
    }
  >
    <Input
      disabled={true}
      style={
        Array [
          Object {
            "color": "#000",
            "flex": 1,
            "fontSize": 17,
            "height": 50,
            "paddingLeft": 5,
            "paddingRight": 5,
          },
          Object {
            "border": "solid 1px red",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        disabled={true}
        editable={false}
        placeholderTextColor="#575757"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "color": "#000",
              "flex": 1,
              "fontSize": 17,
              "height": 50,
              "paddingLeft": 5,
              "paddingRight": 5,
            },
            Object {
              "border": "solid 1px red",
            },
          ]
        }
        underlineColorAndroid="rgba(0,0,0,0)"
      >
        <TextInput
          allowFontScaling={true}
          disabled={true}
          editable={false}
          placeholderTextColor="#575757"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "color": "#000",
                "flex": 1,
                "fontSize": 17,
                "height": 50,
                "paddingLeft": 5,
                "paddingRight": 5,
              },
              Object {
                "border": "solid 1px red",
              },
            ]
          }
          underlineColorAndroid="rgba(0,0,0,0)"
        />
      </TextInput>
    </Input>
  </Styled(Input)>
  <Styled(Styled(Icon))
    name="checkmark-circle"
    status="valid"
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "WrappedComponent": [Function],
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "hUzqNt",
            "rules": Array [
              "
  color: ",
              [Function],
              ";
",
            ],
          },
          "displayName": "Styled(Styled(Icon))",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": [Function],
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      name="checkmark-circle"
      status="valid"
    >
      <Styled(Icon)
        className="sc-bdVaJa hUzqNt"
        name="checkmark-circle"
        status="valid"
      >
        <Icon
          className="sc-bdVaJa hUzqNt"
          name="checkmark-circle"
          status="valid"
          style={
            Object {
              "color": "#000",
              "fontSize": 30,
            }
          }
        >
          <Styled(IconNB)
            className="sc-bdVaJa hUzqNt"
            name="ios-checkmark-circle"
            status="valid"
            style={
              Object {
                "color": "#000",
                "fontSize": 30,
              }
            }
          >
            <IconNB
              className="sc-bdVaJa hUzqNt"
              name="ios-checkmark-circle"
              status="valid"
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                  Object {
                    "color": "#000",
                    "fontSize": 30,
                  },
                ]
              }
            >
              <Icon
                allowFontScaling={false}
                className="sc-bdVaJa hUzqNt"
                name="ios-checkmark-circle"
                size={12}
                status="valid"
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontSize": 30,
                    },
                    Object {
                      "color": "#000",
                      "fontSize": 30,
                    },
                  ]
                }
              >
                <Text
                  allowFontScaling={false}
                  className="sc-bdVaJa hUzqNt"
                  status="valid"
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 12,
                      },
                      Array [
                        Object {
                          "color": "#000",
                          "fontSize": 30,
                        },
                        Object {
                          "color": "#000",
                          "fontSize": 30,
                        },
                      ],
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  <Text
                    allowFontScaling={false}
                    className="sc-bdVaJa hUzqNt"
                    status="valid"
                    style={
                      Array [
                        Object {
                          "color": undefined,
                          "fontSize": 12,
                        },
                        Array [
                          Object {
                            "color": "#000",
                            "fontSize": 30,
                          },
                          Object {
                            "color": "#000",
                            "fontSize": 30,
                          },
                        ],
                        Object {
                          "fontFamily": "Ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                  >
                    ï…Š
                  </Text>
                </Text>
              </Icon>
            </IconNB>
          </Styled(IconNB)>
        </Icon>
      </Styled(Icon)>
    </StyledComponent>
  </Styled(Styled(Icon))>
</TextInput>
`;
